service/subresource_destroyer.go

```diff
@@ -0,0 +1,165 @@
+			}
+		}
+
+		// Raise events for the deleted endpoints.
+		var endpointIds []int64
+		for _, endpoint := range endpoints {
+			endpointIds = append(endpointIds, endpoint.ID)
+			err := RaiseEvent("Endpoint.delete", &endpoint, headers)
+			if err != nil {
+				logging.Log.Errorf(`Event "Endpoint.delete" could not be raised for endpoint %v: %s`, endpoint.ToEvent(), err)
+			}
+		}
+
+		// Raise events for the deleted connections.
+		for _, connection := range rhcConnections {
+			err := RaiseEvent("RhcConnection.delete", &connection, headers)
+			if err != nil {
+				logging.Log.Errorf(`Event "RhcConnection.delete" could not be raised for rhcConnection %v: %s`, connection.ToEvent(), err)
+			}
+		}
+
+		// Raise an event for the source itself.
+		err = RaiseEvent("Source.delete", source, headers)
+		if err != nil {
+			logging.Log.Errorf(`Event "Source.delete" could not be raised for source %v: %s`, source.ToEvent(), err)
+		}
+
+		// Fetch all the authentications from the resources.
+		resourceTypes := []struct {
+			ResourceType string
+			ResourceIds  []int64
+		}{
+			{ResourceType: "Application", ResourceIds: appIds},
+			{ResourceType: "Endpoint", ResourceIds: endpointIds},
+			{ResourceType: "Source", ResourceIds: []int64{resourceId}},
+		}
+
+		for _, res := range resourceTypes {
+			auths, err := authenticationsDao.ListIdsForResource(res.ResourceType, res.ResourceIds)
+			if err != nil {
+				logging.Log.Errorf(`[resource_type: "%s"][resource_ids: "%v"] Could not fetch authentications: %s`, res.ResourceType, res.ResourceIds, err)
+			}
+
+			authentications = append(authentications, auths...)
+		}
+
+	case "Application":
+		applicationsDao := dao.GetApplicationDao(tenantId)
+		applicationAuthentications, application, err := applicationsDao.DeleteCascade(resourceId)
+		if err != nil {
+			return fmt.Errorf(`could not completely delete the application: %s`, err)
+		}
+
+		// Raise an event for the deleted application authentications.
+		for _, appAuth := range applicationAuthentications {
+			err = RaiseEvent("ApplicationAuthentication.delete", application, headers)
+			if err != nil {
+				logging.Log.Errorf(`Event "ApplicationAuthentication.delete" could not be raised for application authentication %v: %s`, appAuth.ToEvent(), err)
+			}
+		}
+
+		// Raise an event for the deleted application itself.
+		err = RaiseEvent("Application.delete", application, headers)
+		if err != nil {
+			logging.Log.Errorf(`Event "Application.delete" could not be raised for application %v: %s`, application.ToEvent(), err)
+		}
+
+		// Fetch all the application's authentications.
+		auths, err := authenticationsDao.ListIdsForResource("Application", []int64{resourceId})
+		if err != nil {
+			logging.Log.Errorf(`[resource_type: "Application"][resource_id: "%v"] Could not fetch authentications: %s`, resourceId, err)
+		}
+
+		authentications = append(authentications, auths...)
+	case "Endpoint":
```

Comment:
Endpoints can have authentications attached to them - so we should probably add the cascade logic to that DAO as well.

Reply:
The enpoints' authentications already get cleaned up: we call `authenticationsDao.ListIdsForResource("Endpoint", []int64{resourceId})` and then we deal with those authentications outside the `switch`.